### PR TITLE
Adding watch option using dmon

### PR DESCRIPTION
**Pull request by mwestphal**
This add support for an "--watch" cli option where the current file is reloaded automatically if changed on disk.
It relies on https://github.com/septag/dmon for detecting the changes.

In order to support this properly, F3DStarter is reworked with a proper event loops that currently supports:
 - Reloading the current file
 - Rendering

This also adds a test for it on linux/apple/windows

 
Script used:

```
while :
do
  sleep 1
  cp cowlow.vtp a.vtp
  sleep 1
  cp cow.vtp a.vtp
done
```

https://github.com/f3d-app/f3d/assets/3129530/3d4510b3-5303-476e-bba9-0e41394b30d0

Fix: #86

### CODE REVIEW

File: application/F3DStarter.cxx [hunk 1]
```diff
@@ -653,7 +686,9 @@ int F3DStarter::AddFile(const fs::path& path, bool quiet)
 
     if (it == this->Internals->FilesList.end())
     {
+      this->Internals->FilesListMutex.lock();
       this->Internals->FilesList.push_back(tmpPath);
+      this->Internals->FilesListMutex.unlock();
       return static_cast<int>(this->Internals->FilesList.size()) - 1;
```
Comment (snoyer) [inline application/F3DStarter.cxx:692]: get size before unlocking
```suggestion
      this->Internals->FilesList.push_back(tmpPath);
      const size_t fileCount = this->Internals->FilesList.size();
      this->Internals->FilesListMutex.unlock();
      return static_cast<int>(fileCount) - 1;
```

Comment (mwestphal) [reply application/F3DStarter.cxx:692]: line 690 is the only place where FilesList can be modfied. Concurrent read doesnt matter, does it ?

Comment (snoyer) [reply application/F3DStarter.cxx:692]: I don't know, it _feels_ like if you're locking around the push you may want to get the updated size before another push has a chance to get in there.
But I you know for sure that specific piece of code isn't going to be concurrent with itself I guess it's fine either way

Comment (mwestphal) [reply application/F3DStarter.cxx:692]: yes, this mutex is there ONLY for the dmon case. There are many other read of FileLists everywhere in the file, I'd rather not add mutex code there either. I can add a comment though.

Comment (Meakk) [reply application/F3DStarter.cxx:692]: Let's use a `std::lock_guard` anyway

File: application/F3DStarter.cxx [hunk 2]
```diff
@@ -132,15 +137,36 @@ class F3DStarter::F3DInternals
     }
   }
 
+  static void dmonFolderChanged(dmon_watch_id watch_id, dmon_action action, const char* rootdir,
+    const char* filename, const char*, void* userData)
+  {
+    F3DStarter* self = reinterpret_cast<F3DStarter*>(userData);
+    self->Internals->FilesListMutex.lock();
```
Comment (Meakk) [inline application/F3DStarter.cxx:144]: again let's use a scoped lock (`std::lock_guard`)

File: application/F3DStarter.h [hunk 3]
```diff
@@ -34,10 +34,15 @@ class F3DStarter
   void LoadFile(int index = 0, bool relativeIndex = false);
 
   /**
-   * Trigger a render
+   * Trigger a render on the next event loop
    */
   void Render();
 
+  /**
+   * Trigger a render now
```
Comment (Meakk) [inline application/F3DStarter.h:42]:
```suggestion
   * Trigger a render immediately (must be called by the main thread)
```

File: application/F3DStarter.h [hunk 4]
```diff
@@ -47,6 +52,19 @@ class F3DStarter
...
   std::unique_ptr<F3DInternals> Internals;
+
+  /**
+   * Internal method triggered when interacting with the application
+   * that load a file using relative index and handle camera restore
+   */
+  bool LoadRelativeFile(int relativeIndex = 0, bool restoreCamera = false);
+
+  /**
+   * Internal event loop that is triggered repeatdly to handle specific events:
```
Comment (Meakk) [inline application/F3DStarter.h:63]:
```suggestion
   * Internal event loop that is triggered repeatedly to handle specific events:
```

File: application/testing/auto_reload.sh [hunk 5]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
```
Comment (Meakk) [inline application/testing/auto_reload.sh:1]: should we use a Python script instead of bash to have a single script for all OS?

File: application/F3DOptionsParser.cxx [hunk 6]
```diff
@@ -311,6 +311,7 @@ void ConfigurationOptions::GetOptions(F3DAppOptions& appOptions, f3d::options& o
     this->DeclareOption(grp0, "dry-run", "", "Do not read the configuration file", appOptions.DryRun,  HasDefault::YES, MayHaveConfig::NO );
     this->DeclareOption(grp0, "no-render", "", "Do not render anything and quit right after loading the first file, use with --verbose to recover information about a file.", appOptions.NoRender,  HasDefault::YES, MayHaveConfig::YES );
     this->DeclareOption(grp0, "max-size", "", "Maximum size in Mib of a file to load, negative value means unlimited", appOptions.MaxSize,  HasDefault::YES, MayHaveConfig::YES, "<size in Mib>");
+    this->DeclareOption(grp0, "auto-reload", "", "Reload currently loaded file if it is modified on disk", appOptions.AutoReload,  HasDefault::YES, MayHaveConfig::YES );
```
Comment (snoyer) [inline application/F3DOptionsParser.cxx:314]: It looks like "file watching" is the accepted terminology and `--watch` is already a thing out there ([`webpack`](https://webpack.js.org/api/cli/), [`tsc`](https://www.typescriptlang.org/docs/handbook/compiler-options.html), [`spawning(1)`](https://linux.die.net/man/1/spawning), [`qiv`](https://linux.die.net/man/1/qiv), [`kubectl`](https://kubernetes.io/docs/reference/kubectl/), ...)

For the description I'm not 100% but maybe using "whenever" instead of "if" is enough to convey the automatic/continuous/repeating idea?
```suggestion
    this->DeclareOption(grp0, "watch", "", "Reload the current file whenever it is modified on disk", appOptions.AutoReload,  HasDefault::YES, MayHaveConfig::YES );
```
also "currently loaded file" vs "current file" vs "active file"? idk

File: application/testing/auto_reload.sh (hunk 5 above)
Comment (mwestphal) [reply application/testing/auto_reload.sh:1]: i'd rather write a powershell script instead of adding a dependency to python in our testing.

File: application/testing/test_watch.sh [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+sleep 1
+cp $lowres_data $reloaded_data
+sleep 1
+
+grep -q "Number of points: 634" $log
+ret=$?
+
+kill $pid
+exit $ret
```
Comment (mwestphal) [inline application/testing/test_watch.sh:30]: need a trap to improve the test

File: application/testing/test_watch.ps1 [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+$str = Select-String $log -Pattern "Number of points: 634"
+if ($str -ne $null) {
+  $ret = 0
+} else {
+  $ret = 1
+}
+
+Stop-Process -Id $id
+exit $ret
```
Comment (mwestphal) [inline application/testing/test_watch.ps1:30]: need a trap to improve the script

File: application/testing/CMakeLists.txt [hunk 9]
```diff
@@ -897,3 +897,14 @@ if(NOT F3D_MACOS_BUNDLE)
...
   endif()
 endif()
+
+if(NOT VTK_OPENGL_HAS_EGL)
+  if(UNIX)
+    # Custom bash test for testing watch option
+    add_test (NAME f3d::TestWatch COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/test_watch.sh $<TARGET_FILE:f3d> ${F3D_SOURCE_DIR}/testing/data ${CMAKE_BINARY_DIR}/Testing/Temporary)
+  elseif(WIN32)
+    find_program(PWSH NAMES pwsh powershell)
+    add_test (NAME f3d::TestWatch COMMAND ${PWSH} ${CMAKE_CURRENT_SOURCE_DIR}/test_watch.ps1 $<TARGET_FILE:f3d> ${F3D_SOURCE_DIR}/testing/data ${CMAKE_BINARY_DIR}/Testing/Temporary)
```
Comment (mwestphal) [inline application/testing/CMakeLists.txt:907]: could be improved with some cmake logic

File: application/F3DStarter.cxx [hunk 10]
```diff
@@ -6,15 +6,29 @@
 #include "F3DOptionsParser.h"
 #include "F3DSystemTools.h"
 
+#define DMON_IMPL
+#ifdef WIN32
+#pragma warning(push)
+#pragma warning(disable : 4505)
+#include "dmon.h"
+// dmon includes Windows.h which defines 'ERROR' and conflicts with log.h
```
Comment (Meakk) [inline application/F3DStarter.cxx:14]: Can you try to put `#define WIN32_LEAN_AND_MEAN` before `#include "dmon.h"` to see if `ERROR` is still defined?

**Comment by snoyer** [comment]
Looks like you already have the implementation details covered so I'm left with nitpicking naming:

- `ForceRender()` just performs the render it doesn't take any extra steps to "force" it -> `Render()`
- `Render()` doens't actually perform the render, it only asks for the render to be done (or declares the current render outdated depending on how you want to look at it) -> `RequestRender()`, `InvalidateRender()`, ...
- `SoftRender` is a flag but is named like an action -> `RenderRequested`, `RenderNeeded`, `RenderFlag`, ... (same for `ReloadFile` -> `ReloadRequested`, ...)

tldr:

```cpp
void Render() { SoftRender = true; } // do the thing = don't actually  o_O
if (SoftRender) ForceRender(); // if soft then forceful  o_O
```
```cpp
void RequestRender() { RenderRequested = true; } // request the thing = mark it as requested  ^_^
if (RenderRequested) Render(); // do the thing if asked  ^_^
```

**Comment by mwestphal** [comment]
As always, the easiest to find out how to do the right thing is just to do the wrong thing and wait for people to fix it for you, thanks @snoyer :)

File: application/F3DStarter.cxx (hunk 10 above)
Comment (mwestphal) [reply application/F3DStarter.cxx:15]: does not work sadly.

File: application/testing/test_watch.sh [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+log=$tmp_dir/output.log
+$f3d_cmd --watch --verbose $reloaded_data > $log &
+pid=$!
+
+function cleanup()
+{
+  if [[ -z "${CTEST_F3D_COVERAGE-}" ]]; then
+    kill $pid
+  else
+    xdotool key Escape
```
Comment (mwestphal) [inline application/testing/test_watch.sh:27]: @Meakk should I simplify this and require xdotool for testing all the time except on only when enabling F3D_COVERAGE ?

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -257,7 +257,7 @@ jobs:
     if: github.event.pull_request.draft == false
 
     runs-on: ubuntu-latest
-    container: ghcr.io/f3d-app/f3d-ci
+    container: mwestphal/f3d-ci
```
Comment (Meakk) [inline .github/workflows/ci.yml:260]: to change

File: application/testing/test_watch.sh (hunk 11 above)
Comment (Meakk) [reply application/testing/test_watch.sh:27]: The problem is `xdotool` won't work on macos.

Comment (mwestphal) [reply application/testing/test_watch.sh:27]: very good point, lets keep as is.

File: application/testing/test_watch.ps1 (hunk 8 above)
Comment (mwestphal) [reply application/testing/test_watch.ps1:30]: lets not, its fine as is.

File: application/testing/CMakeLists.txt [hunk 13]
```diff
@@ -899,3 +899,20 @@ if(NOT F3D_MACOS_BUNDLE)
     f3d_test(NAME TestScanPluginsCheckExodus ARGS --scan-plugins NO_RENDER NO_BASELINE REGEXP " - exodus")
   endif()
 endif()
+
+# Watch testing require onscreen renderong
```
Comment (Meakk) [inline application/testing/CMakeLists.txt:903]:
```suggestion
# Watch testing require onscreen rendering
```

**Review by Meakk — APPROVED** [review]
small typo to fix and good to go!
